File: modules/agent/context.go
package agent

import (
	"github.com/woyin/OrangeCoding/modules/core"
)

// AgentContext holds the conversation state, environment, and metadata for an agent session.
type AgentContext struct {
	sessionID    core.SessionId
	conversation *core.Conversation
	workDir      string
	env          map[string]string
	metadata     map[string]string
}

// NewAgentContext creates a new AgentContext with the given session ID and working directory.
func NewAgentContext(sessionID core.SessionId, workDir string) *AgentContext {
	return &AgentContext{
		sessionID:    sessionID,
		conversation: core.NewConversation(),
		workDir:      workDir,
		env:          make(map[string]string),
		metadata:     make(map[string]string),
	}
}

// SetSystemPrompt sets the system prompt as the first message in the conversation.
// If the conversation already has a system message, it is replaced.
func (c *AgentContext) SetSystemPrompt(prompt string) {
	msgs := c.conversation.Messages()
	if len(msgs) > 0 && msgs[0].Role == core.RoleSystem {
		// Rebuild conversation with new system prompt
		newConv := core.NewConversation()
		newConv.AddMessage(core.NewSystemMessage(prompt))
		for _, m := range msgs[1:] {
			newConv.AddMessage(m)
		}
		c.conversation = newConv
	} else {
		// Prepend system message by rebuilding
		newConv := core.NewConversation()
		newConv.AddMessage(core.NewSystemMessage(prompt))
		for _, m := range msgs {
			newConv.AddMessage(m)
		}
		c.conversation = newConv
	}
}

// AddUserMessage appends a user message to the conversation.
func (c *AgentContext) AddUserMessage(content string) {
	c.conversation.AddMessage(core.NewUserMessage(content))
}

// AddAssistantMessage appends an assistant message to the conversation.
func (c *AgentContext) AddAssistantMessage(content string) {
	c.conversation.AddMessage(core.NewAssistantMessage(content))
}

// AddToolResult appends a tool result message to the conversation.
func (c *AgentContext) AddToolResult(result core.ToolResult) {
	c.conversation.AddMessage(result.ToMessage())
}

// Conversation returns the underlying conversation.
func (c *AgentContext) Conversation() *core.Conversation {
	return c.conversation
}

// SessionID returns the session ID.
func (c *AgentContext) SessionID() core.SessionId {
	return c.sessionID
}

// WorkDir returns the working directory.
func (c *AgentContext) WorkDir() string {
	return c.workDir
}

File: packages/agent/src/context.ts
/**
 * AgentContext holds the conversation state, environment, and metadata for an agent session.
 * Ported from modules/agent/context.go.
 */

import {
  Role,
  SessionId,
  Conversation,
  Message,
  newSystemMessage,
  newUserMessage,
  newAssistantMessage,
  newToolResultMessage,
} from "@orangecoding/core";
import type { ToolResult, ToolCall } from "@orangecoding/core";
import type { HarnessProfile } from "./harness-profile.js";

export class AgentContext {
  private _sessionID: SessionId;
  private _conversation: Conversation;
  private _workDir: string;
  private _env: Map<string, string>;
  private _metadata: Map<string, string>;

  constructor(sessionID: SessionId, workDir: string) {
    this._sessionID = sessionID;
    this._conversation = Conversation.create();
    this._workDir = workDir;
    this._env = new Map();
    this._metadata = new Map();
  }

  /** SetSystemPrompt sets the system prompt as the first message in the conversation. */
  setSystemPrompt(prompt: string): void {
    this._conversation.replaceSystemPrompt(prompt);
  }

  /** ApplyHarnessProfile appends harness guidance to the system prompt once. */
  applyHarnessProfile(profile: HarnessProfile): void {
    const addendum = profile.systemPromptAddendum();
    const current = this._conversation.systemPrompt();
    if (current === undefined) {
      this.setSystemPrompt(addendum.trim());
      return;
    }
    if (current.includes("[OrangeCoding Harness]")) {
      return;
    }
    this.setSystemPrompt(current + addendum);
  }

  /** AddUserMessage appends a user message to the conversation. */
  addUserMessage(content: string): void {
    this._conversation.addMessage(newUserMessage(content));
  }

  /** AddAssistantMessage appends an assistant message to the conversation. */
  addAssistantMessage(content: string): void {
    this._conversation.addMessage(newAssistantMessage(content));
  }

  /** AddToolResult appends a tool result message to the conversation. */
  addToolResult(result: ToolResult): void {
    this._conversation.addMessage(result.toMessage());
  }

  /** Conversation returns the underlying conversation. */
  get conversation(): Conversation {
    return this._conversation;
  }

  /** SessionID returns the session ID. */
  get sessionID(): SessionId {
    return this._sessionID;
  }

  /** WorkDir returns the working directory. */
  get workDir(): string {
    return this._workDir;
  }
}
